src/mlpack/methods/linear_svm/linear_svm.hpp

```diff
@@ -0,0 +1,195 @@
+   * point. It then chooses the class which has the highest probability among
+   * all.
+   *
+   * @param data Set of points to classify.
+   * @param labels Predicted labels for each point.
+   */
+  void Classify(const MatType& data,
+                arma::Row<size_t>& labels) const;
+
+  /**
+   * Classify the given points, returning class scores and predicted
+   * class label for each point.
+   * The function calculates the scores for every class, given a data
+   * point. It then chooses the class which has the highest probability among
+   * all.
+   *
+   * @param data Matrix of data points to be classified.
+   * @param labels Predicted labels for each point.
+   * @param scores Class probabilities for each point.
+   */
+  void Classify(const MatType& data,
+                arma::Row<size_t>& labels,
+                arma::mat& scores) const;
+
+  /**
+   * Classify the given points, returning class scores for each point.
+   *
+   * @param data Matrix of data points to be classified.
+   * @param scores Class scores for each point.
+   */
+  void Classify(const MatType& data,
+                arma::mat& scores) const;
+
+  /**
+   * Classify the given point. The predicted class label is returned.
+   * The function calculates the scores for every class, given the point.
+   * It then chooses the class which has the highest probability among all.
+   *
+   * @param point Point to be classified.
+   * @return Predicted class label of the point.
+   */
+  template<typename VecType>
+  size_t Classify(const VecType& point) const;
+
+  /**
+   * Computes accuracy of the learned model given the feature data and the
+   * labels associated with each data point. Predictions are made using the
+   * provided data and are compared with the actual labels.
+   *
+   * @param testData Matrix of data points using which predictions are made.
+   * @param testLabels Vector of labels associated with the data.
+   * @return Accuracy of the model.
+   */
+  double ComputeAccuracy(const MatType& testData,
+                         const arma::Row<size_t>& testLabels) const;
+
+  /**
+   * Train the Linear SVM with the given training data.
+   *
+   * @tparam OptimizerType Desired optimizer
+   * @param data Input training features. Each column associate with one sample
+   * @param labels Labels associated with the feature data.
+   * @param numClasses Number of classes for classification.
+   * @param lambda L2-regularization constant.
+   * @param optimizer Desired optimizer.
+   * @return Objective value of the final point.
+   */
+  template <typename OptimizerType>
+  double Train(const MatType& data,
+               const arma::Row<size_t>& labels,
+               const size_t numClasses = 2,
+               OptimizerType optimizer = OptimizerType());
+
+
+  //! Sets the number of classes.
+  size_t& NumClasses() { return numClasses; }
+  //! Gets the number of classes.
+  size_t NumClasses() const { return numClasses; }
```

Comment:
Typically for an SVM, the way to do classification for more than two classes is to build `numClasses` all-vs-one classifiers, and then for a test point, to select the class with the greatest margin.  Are you doing something different here?  Or is this restricted to two classes?

Reply:
In my view, either we should restrict `LinearSVMFunction` to two classes and then internally hold multiple SVMs to do multiclass; or, if you are doing something different, we should make it clear what strategy we are using for multiclass linear SVM in the comments.  Let me know if I overlooked anything. :+1:

Reply:
I'm using [Weston and Watkins](https://www.elen.ucl.ac.be/Proceedings/esann/esannpdf/es1999-461.pdf) approach for this problem (which distinguish all classes in one single optimization processing).
Here, I evaluate the loss as
![image](https://user-images.githubusercontent.com/34449856/53539695-9a090300-3b38-11e9-9b63-c748b3240c23.png)
This doesn't restrict `LinearSVMFunction` to two classes. 
So, this implementation designs a single objective function for training all `numClasses`-binary SVMs simultaneously and maximize the margins from each class to the remaining ones.

Reply:
Perfect, thank you for the clarification.  Can you link to that in the documentation somewhere (and possibly document that this is an extension of binary linear SVM to multiclass)?  I'll review the paper soon and then I'll be able to check the implementation fully. :+1:

Reply:
I have linked the document in the comment, but here it is again (https://www.elen.ucl.ac.be/Proceedings/esann/esannpdf/es1999-461.pdf).

Reply:
Ah, now I realize what you mean.  For some reason I didn't notice that "Weston and Watkins" was linked, sorry!